src/main/java/org/opentripplanner/model/StopPattern.java

```diff
@@ -34,14 +34,14 @@
  * A StopPattern is very closely related to a TripPattern -- it essentially serves as the unique
  * key for a TripPattern. Should the route be included in the StopPattern?
  */
-public final class StopPattern implements Serializable {
+public final class StopPattern implements Cloneable, Serializable {
 
     private static final long serialVersionUID = 20140101L;
     public static final int NOT_FOUND = -1;
 
     private final StopLocation[] stops;
-    private final PickDrop[]  pickups;
-    private final PickDrop[]  dropoffs;
+    private PickDrop[]  pickups;
```

Comment:
I'm not sure if this change is appropriate or not. Being able to clone the StopPattern and then just slightly changing the pickups/dropoffs made my changes easier to implement.

Reply:
I do not like it, this breaks the encapsulation - note the `pickups` and `dropoffs` are not visible outside the class. It should be easy to protect the state. Add a new constructor
```java
private StopPattern(StopLocation[] stops, PickDrop[] pickups, PickDrop[] dropoffs) {
    this.stops = stops;
    this.pickups = pickups;
    this.dropoffs = dropoffs;
}
```
And a builder:
```java
public Builder mutate() {
    return new Builder(this);
}
    
static class Builder {
    public final StopLocation[] stops;
    public final PickDrop[]  pickups;
    public final PickDrop[]  dropoffs;

    public Builder(StopPattern original) {
        stops = Arrays.copyOf(original.stops, original.stops.length);
        pickups = Arrays.copyOf(original.pickups, original.pickups.length);
        dropoffs = Arrays.copyOf(original.dropoffs, original.dropoffs.length);
    }
        
    public StopPattern build() {
       return new StopPattern(stops, pickups, dropoffs);
    }
} 
```

Reply:
I agree. There is already too much mutable state in the graph. Lets not make it worse.

Reply:
Could you not do the same as the Siri updater does?

Reply:
Added https://github.com/opentripplanner/OpenTripPlanner/pull/3960/commits/2972c9491f19fd87a05c48541a2bd8d0bc5083d9 that uses the builder approach suggested by @t2gran